src/provider.py

```diff
@@ -0,0 +1,185 @@
+
+        Args:
+            event: the event needing config
+
+        Returns:
+            Dict of `username`, `password` and `endpoints` data for the related app
+        """
+        relation = event.relation
+
+        username = f"relation-{relation.id}"
+        password = self.peer_relation.data[self.charm.app].get(username) or generate_password()
+        bootstrap_server = self.charm.kafka_config.bootstrap_server
+        endpoints = [server.split(":")[0] for server in bootstrap_server]
+        zookeeper_uris = self.charm.kafka_config.zookeeper_config.get("connect", "")
+
+        relation_config = {
+            "username": username,
+            "password": password,
+            "endpoints": ",".join(endpoints),
+            "uris": ",".join(bootstrap_server),
+            "zookeeper-uris": zookeeper_uris,
+            "consumer-group-prefix": "",
+        }
+
+        # only set this if `consumer` is set to avoid missing information
+        if "consumer" in event.relation.data[event.app].get("extra-user-roles", ""):
+            relation_config["consumer-group-prefix"] = f"{username}-"
+
+        return relation_config
+
+    def update_acls(self, event: RelationChangedEvent) -> None:
+        """Updates cluster ACLs for a given event relation to match client relation data.
+
+        Args:
+            event: the event from a related client application needing ACLs
+        """
+        if not self.charm.unit.is_leader():
+            return
+
+        if not self.charm.kafka_config.zookeeper_connected:
+            logger.debug("cannot update ACLs, ZooKeeper not yet connected")
+            event.defer()
+            return
+
+        provider_relation_config = self.provider_relation_config(event=event)
+        requirer_relation_config = self.requirer_relation_config(event=event)
+
+        self.kafka_auth.load_current_acls()
+
+        self.kafka_auth.update_user_acls(
+            username=provider_relation_config["username"],
+            group=provider_relation_config.get("consumer-group-prefix"),
+            **requirer_relation_config,
+        )
+
+        # non-leader units need cluster_config_changed event to update their super.users
+        self.charm.model.get_relation(PEER).data[self.charm.app].update(
+            {"super-users": self.kafka_config.super_users}
+        )
+
+        event.relation.data[self.charm.app].update(provider_relation_config)
+
+    def _on_relation_created(self, event: RelationCreatedEvent) -> None:
+        """Handler for `kafka-client-relation-created` event.
+
+        Adds new relation users to ZooKeeper.
+
+        Args:
+            event: the event from a related client application needing a user
+        """
+        if not self.charm.unit.is_leader():
+            return
+
+        if not self.charm.ready_to_start:
+            logger.debug("cannot add user, ZooKeeper not yet connected")
+            event.defer()
+            return
```

Comment:
I assume you defer since you're waiting for one of this condition to become true:
```
self.charm.ready_to_start
```
I didn't check this super thoroughly but if in order for `self.charm.ready_to_start` to be true some other events need to run (ie the start/config/install hook) deferring here can cause this event to continuously defer making it take much longer for the start hook to get on the queue before this event re-queues. This can lead to downtime/poor performance. Instead I suggest you set a flag and trigger and event that will re-call `_on_relation_created `.

Mykola did something similar in his k8s charm [here](https://github.com/delgod/mongodb-k8s-operator/blob/ad178a7e2406f3b805ac73d5dfb7927b123c6deb/lib/charms/mongodb_libs/v0/mongodb_provider.py#L60-L63) and instead of re-triggering the event, he creating a function for himself to call once his condition was met by the start hook, see [here](https://github.com/delgod/mongodb-k8s-operator/blob/ad178a7e2406f3b805ac73d5dfb7927b123c6deb/src/charm.py#L142)